compiler/p/codegen/OMRTreeEvaluator.cpp

```diff
@@ -5189,6 +5191,232 @@ TR::Register *OMR::Power::TreeEvaluator::directCallEvaluator(TR::Node *node, TR:
+   //                          but Hursley advised to be conservative due to lack of specification.
+   generateInstruction(cg, TR::InstOpCode::lwsync, node);
+
+   TR::LabelSymbol *doneLabel = TR::LabelSymbol::create(cg->trHeapMemory(),cg);
+   TR::LabelSymbol *loopLabel = TR::LabelSymbol::create(cg->trHeapMemory(),cg);
+
+   loopLabel->setStartInternalControlFlow();
+   deltaChild = node->getSecondChild();
+
+   if (deltaChild->getOpCode().isLoadConst()
+          && !deltaChild->getRegister()
+          && deltaChild->getDataType() == TR::Int32)
+      {
+      delta = (int32_t)(deltaChild->getInt());
+      isDeltaConstant = true;
+
+      //determine if the constant can be represented as an immediate
+      if (delta <= UPPER_IMMED && delta >= LOWER_IMMED)
+         {
+         // avoid evaluating immediates for add operations
+         isDeltaImmediate = true;
+         }
+      else if (delta & 0xFFFF == 0 && (delta & 0xFFFF0000) >> 16 <= UPPER_IMMED && (delta & 0xFFFF0000) >> 16 >= LOWER_IMMED)
+         {
+         // avoid evaluating shifted immediates for add operations
+         isDeltaImmediate = true;
+         isDeltaImmediateShifted = true;
+         }
+      else
+         {
+         // evaluate non-immediate constants since there may be reuse
+         // and they have to go into a reg anyway
+         tempReg = cg->evaluate(deltaChild);
+         }
+      }
+   else
+      {
+      tempReg = cg->evaluate(deltaChild);
+      }
+
+   generateLabelInstruction(cg, TR::InstOpCode::label, node, loopLabel);
+
+   deltaReg = cg->allocateRegister();
+   if (isDeltaImmediate && !isAddOp)
+      {
+      // If argument is immediate, but the operation is not an add,
+      // the value must still be loaded into a register
+      // If argument is an immediate value and operation is an add,
+      // it will be used as an immediate operand in an add immediate instruction
+      loadConstant(cg, node, delta, deltaReg);
+      }
+   else if (!isDeltaImmediate)
+      {
+      // For non-constant arguments, use evaluated register
+      // For non-immediate constants, evaluate since they may be re-used
+      numDeps++;
+      generateTrg1Src1Instruction(cg, TR::InstOpCode::mr, node, deltaReg, tempReg);
+      }
+
+   uint8_t len = isLong ? 8 : 4;
+
+   generateTrg1MemInstruction(cg, isLong ? TR::InstOpCode::ldarx : TR::InstOpCode::lwarx, node, resultReg,
+         new (cg->trHeapMemory()) TR::MemoryReference(0, valueAddrReg, len, cg));
+
+   if (isAddOp)
+      {
+      if (isDeltaImmediateShifted)
+         generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::addis, node, deltaReg, resultReg, ((delta & 0xFFFF0000) >> 16));
+      else if (isDeltaImmediate)
+         generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::addi, node, deltaReg, resultReg, delta);
+      else
+         generateTrg1Src2Instruction(cg, TR::InstOpCode::add, node, deltaReg, resultReg, deltaReg);
+      }
+
```

Comment:
Maybe I misunderstood it, but AddAndFetch result should be the sum?  Here, you have the loaded value as the result always.

Reply:
@zl-wang Sorry for the confusion, Julian.  At lines #5338-5352, the code will return `resultReg` for `FetchAndAdd` and `deltaReg` for `AddAndFetch` - that's based on the value of `isGetThenUpdate`.  Would it help if I changed the name of `resultReg` to `prevValueReg`, say, and renamed `deltaReg` to `updatedValueReg`?

Reply:
Fixed up variable names to clarify what they hold.  Revised in commit https://github.com/eclipse/omr/pull/3018/commits/e8493aa681a66b85e9d83cca9cbe1f7dfdda4c1b